orchestrator/orchestrator.go

```diff
@@ -0,0 +1,213 @@
+
+	OpSimInstances []*op_simulator.OpSimulator
+	anvilInstances []*anvil.Anvil
+}
+
+//go:embed genesisstates/genesis-l1.json
+var genesisL1JSON []byte
+
+//go:embed genesisstates/genesis-l2.json
+var genesisL2JSON []byte
+
+func NewOrchestrator(log log.Logger, config *OrchestratorConfig) (*Orchestrator, error) {
+	var opSimInstances []*op_simulator.OpSimulator
+	var anvilInstances []*anvil.Anvil
+
+	l1Count := 0
+	for _, config := range config.ChainConfigs {
+		if config.BaseChainId == 0 {
+			l1Count++
+		}
+	}
+
+	if l1Count > 1 {
+		return nil, fmt.Errorf("supersim does not support more than one l1")
+	}
+
+	for _, chainConfig := range config.ChainConfigs {
+		var genesis []byte
+		if chainConfig.BaseChainId == 0 {
+			genesis = genesisL1JSON
+		} else {
+			genesis = genesisL2JSON
+		}
+		anvil := anvil.New(log, &anvil.Config{ChainId: chainConfig.ChainId, Genesis: genesis})
+		anvilInstances = append(anvilInstances, anvil)
+		opSimInstances = append(opSimInstances, op_simulator.New(log, &op_simulator.Config{Port: chainConfig.Port, BaseChainId: chainConfig.BaseChainId}, anvil))
+	}
+
+	return &Orchestrator{log, opSimInstances, anvilInstances}, nil
+}
+
+func (o *Orchestrator) Start(ctx context.Context) error {
+	o.log.Info("starting orchestrator")
+
+	for _, anvilInstance := range o.anvilInstances {
+		if err := anvilInstance.Start(ctx); err != nil {
+			return fmt.Errorf("anvil instance chain.id=%v failed to start: %w", anvilInstance.ChainId(), err)
+		}
+	}
+	for _, opSimInstance := range o.OpSimInstances {
+		if err := opSimInstance.Start(ctx); err != nil {
+			return fmt.Errorf("op simulator instance chain.id=%v failed to start: %w", opSimInstance.ChainId(), err)
+		}
+	}
+
+	if err := o.WaitUntilReady(); err != nil {
+		return fmt.Errorf("orchestrator failed to get ready: %w", err)
+	}
+
+	o.enableAnvilLogging()
+
+	o.log.Info("orchestrator is ready")
+
+	return nil
+}
+
+func (o *Orchestrator) Stop(ctx context.Context) error {
+	o.log.Info("stopping orchestrator")
+
+	for _, opSim := range o.OpSimInstances {
+		if err := opSim.Stop(ctx); err != nil {
+			return fmt.Errorf("op simulator chain.id=%v failed to stop: %w", opSim.ChainId(), err)
+		}
+		o.log.Info("stopped op simulator", "chain.id", opSim.ChainId())
+	}
+	for _, anvil := range o.anvilInstances {
+		if err := anvil.Stop(); err != nil {
+			return fmt.Errorf("anvil chain.id=%v failed to stop: %w", anvil.ChainId(), err)
+		}
+		o.log.Info("stopped anvil", "chain.id", anvil.ChainId())
+	}
+
+	o.log.Info("stopped orchestrator")
+
+	return nil
+}
+
+func (o *Orchestrator) Stopped() bool {
+	for _, anvil := range o.anvilInstances {
+		if stopped := anvil.Stopped(); !stopped {
+			return stopped
+		}
+	}
+	for _, opSim := range o.OpSimInstances {
+		if stopped := opSim.Stopped(); !stopped {
+			return stopped
+		}
+	}
+
+	return true
+}
+
+func (o *Orchestrator) WaitUntilReady() error {
+	var once sync.Once
+	var err error
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
I'm wondering if we should consider adding a timeout at all https://pkg.go.dev/context#WithTimeout

Reply:
@nitaliano I think for now we dont need one since this is only waiting on anvil instances to be ready, and that wait function already has a timeout in it: https://github.com/ethereum-optimism/supersim/blob/e556e6e296dea56a6ef5b4b23ae4d250477a5c07/anvil/anvil.go#L204. In the future it could be necessary if there are other services we need to wait on. I'll leave it out for now since we don't currently need it. Let me know if you think otherwise!